File: packages/actant-mcp-server/src/predicate.ts
/**
 * Predicate evaluator — JS port of `crates/actant-subscribe/src/predicate.rs`.
 *
 * Predicates are tagged objects: `{ "op": "eq", "field": "...", "value": ... }`.
 * Logical ops: `{ "op": "and", "args": [...] }`, `or`, `not`.
 * Comparators: eq, ne, lt, le, gt, ge, exists.
 * Special: `{ "op": "true" }` / `{ "op": "false" }`.
 *
 * Dotted paths with numeric segments index into arrays (`items.0.id`).
 *
 * No new public type — predicate JSON is accepted at the tool boundary
 * and parsed here. Source of truth for the shape lives in the Rust crate.
 */

export type JSONValue =
  | null
  | boolean
  | number
  | string
  | JSONValue[]
  | { [k: string]: JSONValue };

export type Predicate =
  | { op: "true" }
  | { op: "false" }
  | { op: "eq"; field: string; value: JSONValue }
  | { op: "ne"; field: string; value: JSONValue }
  | { op: "lt"; field: string; value: JSONValue }
  | { op: "le"; field: string; value: JSONValue }
  | { op: "gt"; field: string; value: JSONValue }
  | { op: "ge"; field: string; value: JSONValue }
  | { op: "exists"; field: string }
  | { op: "and"; args: Predicate[] }
  | { op: "or"; args: Predicate[] }
  | { op: "not"; arg: Predicate };

export function evaluatePredicate(pred: unknown, root: JSONValue): boolean {
  if (!isObject(pred)) return false;
  const op = (pred as { op?: unknown }).op;
  switch (op) {
    case "true":
      return true;
    case "false":
      return false;
    case "exists":
      return resolve(root, (pred as { field: string }).field) !== undefined;
    case "eq": {
      const v = resolve(root, (pred as { field: string }).field);
      if (v === undefined) return false;
      return deepEq(v, (pred as { value: JSONValue }).value);
    }
    case "ne": {
      const v = resolve(root, (pred as { field: string }).field);
      // Documented exception in the Rust predicate: missing => true for `ne`.
      if (v === undefined) return true;
      return !deepEq(v, (pred as { value: JSONValue }).value);
    }
    case "lt":
      return cmp(pred, root, (o) => o < 0);
    case "le":
      return cmp(pred, root, (o) => o <= 0);
    case "gt":
      return cmp(pred, root, (o) => o > 0);
    case "ge":
      return cmp(pred, root, (o) => o >= 0);
    case "and": {
      const xs = (pred as { args?: unknown }).args;
      if (!Array.isArray(xs)) return false;
      return xs.every((p) => evaluatePredicate(p, root));
    }
    case "or": {
      const xs = (pred as { args?: unknown }).args;
      if (!Array.isArray(xs)) return false;
      return xs.some((p) => evaluatePredicate(p, root));
    }
    case "not":
      return !evaluatePredicate((pred as { arg: Predicate }).arg, root);
    default:
      return false;
  }
}

function cmp(pred: unknown, root: JSONValue, ok: (n: number) => boolean): boolean {
  const field = (pred as { field?: string }).field;
  const lit = (pred as { value?: JSONValue }).value;
  if (typeof field !== "string" || lit === undefined) return false;
  const got = resolve(root, field);
  if (got === undefined) return false;
  const ord = order(got, lit);
  if (ord === undefined) return false;
  return ok(ord);
}

function resolve(root: JSONValue, path: string): JSONValue | undefined {
  if (path.length === 0) return undefined;
  let cur: JSONValue | undefined = root;
  for (const seg of path.split(".")) {
    if (seg.length === 0) return undefined;
    if (cur === null || cur === undefined) return undefined;
    if (Array.isArray(cur)) {
      const idx = Number(seg);
      if (!Number.isInteger(idx) || idx < 0) return undefined;
      cur = cur[idx];
    } else if (typeof cur === "object") {
      cur = (cur as { [k: string]: JSONValue })[seg];
    } else {
      return undefined;
    }
  }
  return cur;
}

function order(a: JSONValue, b: JSONValue): number | undefined {
  if (typeof a === "number" && typeof b === "number") {
    return a - b;
  }
  if (typeof a === "string" && typeof b === "string") {
    return a < b ? -1 : a > b ? 1 : 0;
  }
  if (typeof a === "boolean" && typeof b === "boolean") {
    return Number(a) - Number(b);
  }
  if (a === null && b === null) return 0;
  return undefined;
}

function deepEq(a: JSONValue, b: JSONValue): boolean {
  if (a === b) return true;
  if (a === null || b === null) return false;
  if (typeof a !== typeof b) return false;
  if (Array.isArray(a) || Array.isArray(b)) return arraysDeepEq(a, b);
  if (jsonRecord(a) || jsonRecord(b)) return recordsDeepEq(a, b);
  return false;
}

function arraysDeepEq(a: JSONValue, b: JSONValue): boolean {
  if (!Array.isArray(a) || !Array.isArray(b)) return false;
  if (a.length !== b.length) return false;
  return a.every((value, index) => {
    const other = b[index];
    return other !== undefined && deepEq(value, other);
  });
}

function jsonRecord(value: JSONValue): value is { [k: string]: JSONValue } {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function recordsDeepEq(a: JSONValue, b: JSONValue): boolean {
  if (!jsonRecord(a) || !jsonRecord(b)) return false;
  const ak = Object.keys(a);
  const bk = Object.keys(b);
  if (ak.length !== bk.length) return false;
  return ak.every((k) => {
    const av = a[k];
    const bv = b[k];
    return av !== undefined && bv !== undefined && deepEq(av, bv);
  });
}

function isObject(x: unknown): x is Record<string, unknown> {
  return typeof x === "object" && x !== null && !Array.isArray(x);
}

File: crates/actant-subscribe/src/predicate.rs
//! Row-level subscription predicates.
//!
//! A predicate is a small expression tree evaluated against the
//! [`Message::payload`](crate::Message) JSON value before the message is
//! handed to a subscriber. The language is intentionally minimal: field
//! references, literals, the six standard comparators, And/Or/Not, and
//! `Exists` for membership.
//!
//! Closes GAPS.md row #20.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Predicate AST.
///
/// `Field` paths are dotted; e.g. `"payload.tool_name"` walks `value["payload"]["tool_name"]`.
/// Array indices are supported as numeric segments: `"items.0.id"`.
///
/// # Evaluation rules
///
/// - Comparators on mismatched types return `false` (no coercion).
/// - Comparing JSON `null` to a non-null is `false` for any comparator
///   except `Ne`, which is `true`.
/// - Missing fields evaluate as `false` for every comparator. Use
///   [`Predicate::Exists`] to distinguish "missing" from "present-and-null".
/// - `Not(missing == X)` is `true` (missing makes the inner `false`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename_all = "snake_case")]
pub enum Predicate {
    /// Always-true predicate. Useful for "no filter" via `Some(Predicate::True)`.
    True,
    /// Always-false predicate.
    False,
    /// Field equals literal.
    Eq {
        /// Dotted JSON path.
        field: String,
        /// Literal JSON value.
        value: Value,
    },
    /// Field not-equal to literal.
    Ne {
        /// Dotted JSON path.
        field: String,
        /// Literal JSON value.
        value: Value,
    },
    /// Field strictly less than literal (numeric or string).
    Lt {
        /// Dotted JSON path.
        field: String,
        /// Literal JSON value.
        value: Value,
    },
    /// Field less-or-equal than literal.
    Le {
        /// Dotted JSON path.
        field: String,
        /// Literal JSON value.
        value: Value,
    },
    /// Field strictly greater than literal.
    Gt {
        /// Dotted JSON path.
        field: String,
        /// Literal JSON value.
        value: Value,
    },
    /// Field greater-or-equal than literal.
    Ge {
        /// Dotted JSON path.
        field: String,
        /// Literal JSON value.
        value: Value,
    },
    /// True iff the dotted path resolves to a value (including `null`).
    Exists {
        /// Dotted JSON path.
        field: String,
    },
    /// Logical AND. Empty vector evaluates to `true`.
    And(Vec<Predicate>),
    /// Logical OR. Empty vector evaluates to `false`.
    Or(Vec<Predicate>),
    /// Logical NOT.
    Not(Box<Predicate>),
}

impl Predicate {
    /// Evaluate the predicate against a JSON value.
    pub fn evaluate(&self, root: &Value) -> bool {
        match self {
            Self::True => true,
            Self::False => false,
            Self::Eq { field, value } => match resolve(root, field) {
                Some(v) => v == value,
                None => false,
            },
            Self::Ne { field, value } => match resolve(root, field) {
                Some(v) => v != value,
                // `Ne` is the documented exception: missing returns true so
                // that `Not(Eq{...})` and `Ne{...}` line up for present fields
                // and a missing field is "not equal to" any literal you name.
                None => true,
            },
            Self::Lt { field, value } => cmp(root, field, value, |o| {
                matches!(o, std::cmp::Ordering::Less)
            }),
            Self::Le { field, value } => cmp(root, field, value, |o| {
                matches!(o, std::cmp::Ordering::Less | std::cmp::Ordering::Equal)
            }),
            Self::Gt { field, value } => cmp(root, field, value, |o| {
                matches!(o, std::cmp::Ordering::Greater)
            }),
            Self::Ge { field, value } => cmp(root, field, value, |o| {
                matches!(o, std::cmp::Ordering::Greater | std::cmp::Ordering::Equal)
            }),
            Self::Exists { field } => resolve(root, field).is_some(),
            Self::And(xs) => xs.iter().all(|p| p.evaluate(root)),
            Self::Or(xs) => xs.iter().any(|p| p.evaluate(root)),
            Self::Not(inner) => !inner.evaluate(root),
        }
    }
}

/// Convenience: top-level evaluator. Equivalent to `predicate.evaluate(root)`.
pub fn evaluate(predicate: &Predicate, root: &Value) -> bool {
    predicate.evaluate(root)
}

/// Walk a dotted path; numeric segments index into arrays.
fn resolve<'a>(root: &'a Value, path: &str) -> Option<&'a Value> {
    let mut cur = root;
    for segment in path.split('.') {
        if segment.is_empty() {
            return None;
        }
        cur = match cur {
            Value::Object(map) => map.get(segment)?,
            Value::Array(arr) => {
                let idx: usize = segment.parse().ok()?;
                arr.get(idx)?
            }
            _ => return None,
        };
    }
    Some(cur)
}

fn cmp<F>(root: &Value, field: &str, lit: &Value, ok: F) -> bool
where
    F: Fn(std::cmp::Ordering) -> bool,
{
    let Some(got) = resolve(root, field) else {
        return false;
    };
    let Some(order) = order(got, lit) else {
        return false;
    };
    ok(order)
}

fn order(a: &Value, b: &Value) -> Option<std::cmp::Ordering> {
    match (a, b) {
        (Value::Number(x), Value::Number(y)) => {
            let xf = x.as_f64()?;
            let yf = y.as_f64()?;
            xf.partial_cmp(&yf)
        }
        (Value::String(x), Value::String(y)) => Some(x.cmp(y)),
        (Value::Bool(x), Value::Bool(y)) => Some(x.cmp(y)),
        (Value::Null, Value::Null) => Some(std::cmp::Ordering::Equal),
        _ => None,
    }
}
